### PR TITLE
ES|QL: Perform a final cleanup of the optimizer rules tests

### LINKED ISSUE #109215: Refactor ESQL optimizer rules
Split the `OptimizerRules` class and associated tests into individual top level classes and single-class test files.

Right now, the `OptimizerRules` class is over 1000 lines long, and has another 1000+ lines in core, and all it does is wrap many small static classes.  Those should all be pulled up to the top level, and have individual test files, so it's clearer what our coverage gaps are.

### PR DESCRIPTION
**Pull request by astefan**
This PR cleans up the optimization rules tests by:
- creating a new test class for `PushDownAndCombineFilters` rule and moving there the tests that specifically call this rule
- removing duplicated tests for `FoldNull` and `PropagateNullables` rules; moving specific tests to their own classes
- using the already existent utils methods in `EsqlTestUtils` and removing the duplicated ones from `LogicalPlanOptimizerTests`

Closes https://github.com/elastic/elasticsearch/issues/109215

### CODE REVIEW

File: x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/EsqlTestUtils.java [hunk 1]
```diff
@@ -668,4 +672,12 @@ static Version randomVersion() {
             default -> throw new IllegalArgumentException();
         };
     }
+
+    public static WildcardLike wildcardLike(Expression left, String exp) {
```
Comment (astefan) [inline x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/EsqlTestUtils.java:676]: Moved from LogicalPlanOptimizerTests.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java [hunk 2]
```diff
@@ -209,9 +172,6 @@
 //@TestLogging(value = "org.elasticsearch.xpack.esql:TRACE", reason = "debug")
 public class LogicalPlanOptimizerTests extends ESTestCase {
 
-    private static final Literal ONE = L(1);
-    private static final Literal TWO = L(2);
-    private static final Literal THREE = L(3);
```
Comment (astefan) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:214]: Reused the ones from EsqlTestUtils.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java [hunk 3]
```diff
@@ -760,102 +720,6 @@ public void testMultipleCombineLimits() {
...
-        return new GreaterThan(EMPTY, left, right, randomZone());
-    }
-
-    public static LessThan lessThanOf(Expression left, Expression right) {
-        return new LessThan(EMPTY, left, right, randomZone());
-    }
-
-    public static GreaterThanOrEqual greaterThanOrEqualOf(Expression left, Expression right) {
-        return new GreaterThanOrEqual(EMPTY, left, right, randomZone());
-    }
```
Comment (astefan) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:773]: These already exist in EsqlTestUtils.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/FoldNullTests.java [hunk 4]
```diff
@@ -34,11 +144,23 @@ public void testNullFoldingIsNull() {
...
         FoldNull rule = new FoldNull();
         // arithmetic
-        assertNullLiteral(rule.rule(new Add(EMPTY, getFieldAttribute(), NULL)));
+        assertNullLiteral(rule.rule(new Add(EMPTY, getFieldAttribute("a"), NULL)));
         // comparison
-        assertNullLiteral(rule.rule(greaterThanOf(getFieldAttribute(), NULL)));
+        assertNullLiteral(rule.rule(greaterThanOf(getFieldAttribute("a"), NULL)));
         // regex
         assertNullLiteral(rule.rule(new RLike(EMPTY, NULL, new RLikePattern("123"))));
+        // date functions
```
Comment (astefan) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/FoldNullTests.java:152]: The testGenericNullableExpression variant from LogicalPlanOptimizerTests was more extensive, thus the additional lines here.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java
Comment (alex-spies) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:1]: Nice job deduplicating all the tests in here and in `PropagateNullableTests`!

**Review by alex-spies — APPROVED** [review]
Nice, thanks @astefan ! Our tests are becoming more easily navigatable :)

I'm not sure all work is done on https://github.com/elastic/elasticsearch/issues/109215, however. While the optimizer rules are split apart, I think the non-local `*OptimizerTests` classes are still huge (6k lines on the logical one, 5k on the physical); ideally, these would also be split apart wherever tests belong mostly to specific optimizer rules, like `testPruneMixedAggInsideUnusedEval` which tests `PruneColumns` (in the context of the whole optimizer).
I think it's fine to call it a day on https://github.com/elastic/elasticsearch/issues/109215 though as long as we try to get the structure of the tests closer to the structure of the optimizer rules as we work on them.

**Comment by astefan** [comment]
Thanks @alex-spies!

> I think the non-local *OptimizerTests classes are still huge (6k lines on the logical one, 5k on the physical); ideally, these would also be split apart wherever tests belong mostly to specific optimizer rules, like testPruneMixedAggInsideUnusedEval which tests PruneColumns (in the context of the whole optimizer).

That's way more work than just creating individual test files for individual rules and move the obvious ones there. Some (many?, not sure) of those tests rely on more than one rule. This needs a fairly deep knowledge of those specific tests to be able to (to a point) find the right *Test file.

https://github.com/elastic/elasticsearch/issues/109215 has some good intentions based on the advantages one gets from finding specific rules tests in specific files when working on that rule, but I want to point out another aspect that, **for me personally**, is more useful when it comes to a larger-class having a bigger number of tests in: more easily finding (and reusing) utils methods that one might not know about and, also, more easily skimming through examples of queries to take inspiration from.

> I think it's fine to call it a day on https://github.com/elastic/elasticsearch/issues/109215 though as long as we try to get the structure of the tests closer to the structure of the optimizer rules as we work on them.

Strongly agree on this one. With time, and with a bit of discipline and willingness, I think we can clean up further the tests and put them in more individual test files as we work in that area.

**Pull request merged**